### PR TITLE
Sequence labeling refactoring

### LINKED ISSUE #2360: Refactoring of Sequence Tagger Class
**Is your feature/enhancement request related to a problem? Please describe.**
Cleaned-up version of the Sequence Tagger class with separate classes for Viterbi and CRF. Adapted to the new Classifier interface.

**Describe the solution you'd like**
-

### PR DESCRIPTION
**Pull request by whoisjones**
Closes #2360.

### CODE REVIEW

**Comment by helpmefindaname** [comment]
@whoisjones is there any status update on this?
if not, do you mind me creating a PR based on this?

**Comment by whoisjones** [comment]
@helpmefindaname currently shifting the sequence labeler below the DefaultClassifier. We still need a parser for previous models, so it still takes some days, but feel free to contribute on this branch.

File: flair/models/sequence_tagger_utils/crf.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        :param tagset_size: number of tag from tag dictionary
+        """
+        super(CRF, self).__init__()
+        self.tagset_size = tagset_size
+        self.emission = torch.nn.Linear(hidden_dim, tagset_size)
+        self.transitions = torch.nn.Parameter(torch.randn(tagset_size, tagset_size))
+        self.transitions.data.zero_()
+        self.to(flair.device)
+
+    def forward(self, features: torch.Tensor) -> torch.Tensor:
```
Comment (alanakbik) [inline flair/models/sequence_tagger_utils/crf.py:25]: should be `torch.tensor` (i.e. Tensor not capitalized)

File: flair/models/sequence_tagger_model.py [hunk 2]
```diff
@@ -1,764 +1,485 @@
...
-
-def log_sum_exp_batch(vecs):
-    maxi = torch.max(vecs, 1)[0]
-    maxi_bc = maxi[:, None].repeat(1, vecs.shape[1])
-    recti_ = torch.log(torch.sum(torch.exp(vecs - maxi_bc), 1))
-    return maxi + recti_
-
+from .sequence_tagger_utils.crf import CRF
+from .sequence_tagger_utils.viterbi import ViterbiLoss, ViterbiDecoder
+from ..datasets import DataLoader, SentenceDataset
```
Comment (alanakbik) [inline flair/models/sequence_tagger_model.py:21]: why these relative module paths? why not `flair.datasets`?

File: flair/models/sequence_tagger_model.py [hunk 3]
```diff
@@ -396,310 +303,217 @@ def forward(self, sentences: List[Sentence]):
...
-        stop = torch.tensor([self.tag_dictionary.get_idx_for_item(STOP_TAG)], device=flair.device)
-        stop = stop[None, :].repeat(tags.shape[0], 1)
+            # make sure its a list
+            if not isinstance(sentences, list):
+                sentences = [sentences]
 
-        pad_start_tags = torch.cat([start, tags], 1)
-        pad_stop_tags = torch.cat([tags, stop], 1)
+            # filter empty sentences
+            if isinstance(sentences[0], Sentence):
```
Comment (helpmefindaname) [inline flair/models/sequence_tagger_model.py:374]: the if check souldn't be required, as sentences is always of type `List[Sentence]` if typing isn't violated

File: flair/models/sequence_tagger_model.py [hunk 4]
```diff
@@ -396,310 +303,217 @@ def forward(self, sentences: List[Sentence]):
...
+                sentences = [sentence for sentence in sentences if len(sentence) > 0]
 
-        for i in range(len(lens_)):
-            pad_stop_tags[i, lens_[i] :] = self.tag_dictionary.get_idx_for_item(STOP_TAG)
+            # order by length
+            reordered_sentences: List[Union[Sentence, str]] = sorted(sentences, key=lambda s: len(s), reverse=True)
```
Comment (helpmefindaname) [inline flair/models/sequence_tagger_model.py:378]: The Union doesn't make sense, as sentences is of type `List[Sentence]` we will always have `reordered_sentences: List[Sentence]` also, I think mypy is able to auto-infer the type of `reordered_sentences` so the typing might not be necessary

File: flair/models/sequence_tagger_model.py (hunk 3 above)
Comment (alanakbik) [reply flair/models/sequence_tagger_model.py:374]: as yes, right! I'll push a correction, thanks!

File: flair/models/sequence_tagger_model.py (hunk 4 above)
Comment (alanakbik) [reply flair/models/sequence_tagger_model.py:377]: also a good point, this is I think a leftover from times when str could also be passed to the predict function!

**Comment by alanakbik** [comment]
@whoisjones thanks a lot for improving this!

**Pull request merged**